students/nDruP/lesson04/mailroom2.py

```diff
@@ -0,0 +1,167 @@
+            break
+
+        donation_amt = input("Enter their Donation Amount >")
+        if donation_amt.lower() == "exit":
+            break
+        donation_amt = float(donation_amt)
+
+        if donor_name not in donor_dict:
+            donor_dict[donor_name] = []
+        donor_dict[donor_name].append(donation_amt)
+        print_divider()
+        message = f"Dearest {donor_name},\n"
+        message += f"Thank you so much for donation of ${donation_amt:.2f}!\n"
+        message += "We will use your donation to create a real living Pokemon."
+        message += f"\nSincerely,\nWe're a Pyramid Scheme & so is {donor_name}"
+        print(message)
+        print_divider()
+
+        break
+    return
+
+
+def create_report():
+    """
+    Print a list of donors sorted by total historical donation amount.
+    Donor Name, Total Given, Num Gifts, Average Gift
+    """
+    sort_donors = sorted(donor_dict.items(), key=sum_2tuple_by2, reverse=True)
+    col_name_list = ["Donor Name", "Total Given", "Gifts", "Avg Donation"]
+    billi_len = 12
+    donor_col_len = max([len(name) for name, donations in sort_donors])
+    donor_col = "{:<" + f"{donor_col_len}" + "}"
+    money_col = "{:>" + f"{billi_len}" + "}"
+
+    header_row = donor_col + "\t" + money_col + "\t{:}" + "\t" + money_col
+    print(header_row.format(*col_name_list))
+    money_col = money_col[:len(money_col)-1] + ".2f}"
+
+    for name, donations in sort_donors:
+        row = donor_col.format(name)
+        row += "\t$ " + money_col.format(sum(donations))
+        row += f"\t {len(donations):>}"
+        row += "\t$ " + money_col.format(sum(donations)/len(donations))
+        print(row)
+    return
+
+
+def create_letters():
+    """
+    Write a full set of letters to each donor to individual files on disk.
+    Go through all donors in donor_dict, generate a thank you letter,
+    write to disk as a text file.
+    """
+    while True:
+        curdate = (datetime.datetime.now()).strftime("%Y_%m_%d")
+        print("Enter an existing directory to place the letters.")
+        print("Current working directory will be used if given directory is " +
+              "invalid")
+        print("Enter exit to return to main menu.")
+        write_dir = input(">")
+        if write_dir.lower() == 'exit':
+            break
+        if not os.path.isdir(write_dir):
+            write_dir = os.getcwd()
+        for donor, donations in donor_dict.items():
+            thank_you = "Dear " + donor + ",\n\t"
+            thank_you += "We are so grateful for your lifetime generosity of "
+            thank_you += f"${sum(donations):.2f}!"
+            thank_you += " You now have our eternal loyalty. Use it wisely."
+            thank_you += "\nSincerely, \n"
+            thank_you += "We're a pyramid scheme and so is " + donor
+            file_name = donor.replace(' ', '_') + "_" + curdate + ".txt"
+            letter = open(write_dir+"/"+file_name, 'w+')
```

Comment:
this is totally correct, for future reference though you will want to use `os.path.join` instead of manually creating full path

read more: https://stackoverflow.com/questions/13944387/why-use-os-path-join-over-string-concatenation